ydb/core/scheme/scheme_tablecell.h

```diff
@@ -10,79 +10,148 @@
+    static constexpr bool CanInline(size_t size) { return size <= MaxInlineSize(); }
+
 private:
-    ui32 DataSize_ : 30;
-    ui32 IsInline_ : 1;
-    ui32 IsNull_   : 1;
-    union {
-        i64 IntVal;
+    // NotInline (low bit):
+    //   0: IsInline() returns true
+    //   1: IsInline() returns false
+    // NotRefValue (high bit):
+    //   0: data is TRefValue
+    //   1: data is TInlineValue
+    // Null values are IsInline() with a nullptr data pointer and zero size
+    // TCell filled with zeroes is always a null value
+    // This bit arrangment allows methods to check a single flag
+    enum EKind : ui8 {
+        // Both bits 0: null value
+        KindNull = 0,
+        // Low bit 1: ref value
+        KindRefValue = 1,
+        // High bit 1: inline value
+        KindInlineValue = 2,
+    };
+
+    struct TRefValue {
         const char* Ptr;
-        double DoubleVal;
-        float FloatVal;
-        char Bytes[8];
+        ui64 Size : 62;
+        ui64 Kind : 2;
+    };
+
+    struct TInlineValue {
+        char Data[15];
+        ui8 Size : 6;
+        ui8 Kind : 2;
+    };
+
+private:
+    union {
+        char Raw[16];
+        TRefValue Ref;
+        TInlineValue Inline;
     };
 
+    constexpr ui8 GetKind() const noexcept {
+        return ui8(Raw[15]) >> 6;
+    }
+
+    constexpr bool IsRefValue() const noexcept {
+        return (GetKind() & 2) == 0;
+    }
+
+    constexpr bool HasInlineFlag() const noexcept {
+        return (GetKind() & 1) == 0;
+    }
+
 public:
     TCell()
         : TCell(nullptr, 0)
     {}
 
     TCell(TArrayRef<const char> ref)
-        : TCell(ref.begin(), ui32(ref.size()))
-    {
-        Y_ABORT_UNLESS(ref.size() < Max<ui32>(), " Too large blob size for TCell");
-    }
-
-    TCell(const char* ptr, ui32 size)
-        : DataSize_(size)
-        , IsInline_(0)
-        , IsNull_(ptr == nullptr)
-        , Ptr(ptr)
-    {
-        Y_DEBUG_ABORT_UNLESS(ptr || size == 0);
-
-        if (CanInline(size)) {
-            IsInline_ = 1;
-            IntVal = 0;
+        : TCell(ref.data(), ref.size())
+    {}
 
+    TCell(const char* ptr, size_t size) {
+        if (!ptr) {
+            Y_DEBUG_ABORT_UNLESS(size == 0);
+            // All zeroes represents the null value
+            ::memset(Raw, 0, 16);
+        } else if (CanInline(size)) {
+            // We use switch with constant size memcpy for better codegen
             switch (size) {
-                case 8: memcpy(&IntVal, ptr, 8); break;
-                case 7: memcpy(&IntVal, ptr, 7); break;
-                case 6: memcpy(&IntVal, ptr, 6); break;
-                case 5: memcpy(&IntVal, ptr, 5); break;
-                case 4: memcpy(&IntVal, ptr, 4); break;
-                case 3: memcpy(&IntVal, ptr, 3); break;
-                case 2: memcpy(&IntVal, ptr, 2); break;
-                case 1: memcpy(&IntVal, ptr, 1); break;
+                case 15: ::memcpy(Inline.Data, ptr, 15); break;
```

Comment:
CanInline: size <= 14 ... switch(size) case 15:

Why does MaxInlineSize 14 but not 15?

Reply:
We have enough space for 15, but the goal here is parity with TUnboxedValuePod, which inlines up to 14. It's better to cover all physically possible cases, in case MaxInlineSize is bumped to 15 later.